Create helper VIs for session management APIs in DC Power Example

subash-suresh opened the pull request:
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-labview/blob/main/CONTRIBUTING.md).

TODO: Check the above box with an 'x' indicating you've read and followed [CONTRIBUTING.md](https://github.com/ni/measurement-services-labview/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?
Currently the reserve session call returns a list of session info and users have to define their own method to get the exact session's information to use the required session on the pins that is needed to be used. A new helper method has been introduced in this PR to allow the users to get the of session information (single/ multiple) related to a specific pin or a set of pins. Additionally, the logic to reserve and unreserve the session has some repeated logic that could be provided as helpers in the examples and this PR does that as well.

This PR is based on the discussions from #207 

### Why should this Pull Request be merged?

This PR does the following,
1. Added a `Get Session Information.vi` from #221.
2. Added a cluster `gRPC Channel Pool` to maintain the discovery service and session management service gRPC ids.
3. Added polymorphic vi implementation of reserve session(s) which in one version returns a single session information and, in another version, it returns multiple sessions information. This uses `Get Session Information.vi` internally. Similarly, a polymorphic vi implementation is also used for unreserve session(s).
4. Added `Initialize DC Power Session.vi` and `Close DC Power Session` helper vis which are used in 'DC Power Specific Logic.vi` which is the main logic used in the measurement logic.

### What testing has been done?

Validated the dc power example by manually running the example and verifying the results.

subash-suresh commented:
The updates of the different files can be found below,
- `GRPC Channel Pool.ctl`
![image](https://github.com/ni/measurementlink-labview/assets/98541596/13375ef3-cb77-442c-b3d8-1fff1b524de4)
-  `Get Session Information.vi` from #221 
![image](https://github.com/ni/measurementlink-labview/assets/98541596/8005df15-4ea5-4909-8a0d-5614f81ae521)
-  `Reserve Session(s)` Polymorphic VI - The logic is similar to #207 , only difference being that grpc ids are taken using a local control instead of measurement context.
  - `Reserve Session- the single session version.
  
![image](https://github.com/ni/measurementlink-labview/assets/98541596/c63b2537-4bd5-4697-987f-518534e6ac4e)

  - `Reserve Session- the multiple session version.
  
![image](https://github.com/ni/measurementlink-labview/assets/98541596/aa3f707d-aa27-4177-80fd-1c5d1d71badb)

-  `Unreserve Session(s)` Polymorphic VI -  The logic is similar to #207, only difference being that grpc ids are taken using a local control instead of measurement context.  
  - `Unreserve Session- the single session version.
  
![image](https://github.com/ni/measurementlink-labview/assets/98541596/e4bab47e-828b-4c70-8fea-579c8908c7f6)

  - `Unreserve Session- the multiple session version.
  
![image](https://github.com/ni/measurementlink-labview/assets/98541596/d2eb0a70-e50a-4024-aa25-dc543f0e8b67)

- `Create DC Power Session.vi` similar to #207 
![image](https://github.com/ni/measurementlink-labview/assets/98541596/6ceba303-fb7c-4d61-9fc2-d53719ed5531)

- `Destroy DC Power Session.vi` similar to #207
![image](https://github.com/ni/measurementlink-labview/assets/98541596/4c283c18-36af-4c88-b1c1-8e29c51d5bde)

- `DC Power Specific Logic` similar to #207
![image](https://github.com/ni/measurementlink-labview/assets/98541596/d214d1e2-c181-4fac-b5fb-95f193fd914b)

Finally with all the above changes the measurement logic looks like below,
![image](https://github.com/ni/measurementlink-labview/assets/98541596/1096e95f-1825-4e8b-a5da-db69fa3e1cfc)

subash-suresh commented:
Is it better to move the gRPC IDs in the measurement context?. Or is the existing implementation in this PR look good?

dixonjoel commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Reserve Session(s).vi:
I'm open to discussion on this, but I like showing the instance icon and polymorphic selector on the Reserve Session(s) poly VI to clarify what's happening and distinguishing the icon with, e.g. "Reserve Single Session" and "Reserve Mult Sessions". On the unreserve session, I think we should have a single icon and not show the selector since it will just auto-adapt. Something like this.
![image](https://github.com/ni/measurementlink-labview/assets/38357562/507b3500-8c6d-4936-852f-ea53ddea3a22)

subash-suresh replied:
I have updated the reserve and unreserve API icons and the poly selector in the measurement logic as shown below,
![Uploading image.png…]()

dixonjoel reviewed (approved):
Approved pending discussion on poly VI icons.

pbirkhol-ni commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Reserve Session(s).vi:
I think this should show the poly selector because it won't auto adapt. It is probably okay to not show the selector for Unreserve Session(s).vi because it will auto adapt.

pbirkhol-ni commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Unreserve Session.vi:
Use Build Array rather than Insert Into Array.

pbirkhol-ni commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Reserve Session.vi:
This should use standard error handling. The logic should be wrapped in a case structure and only execute if there is no error. This is actually true for all subVIs that shouldn't execute on incoming error.

subash-suresh replied:
Updated as suggested

pbirkhol-ni commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Reserve Sessions.vi:
This should use standard error handling. The logic should be wrapped in a case structure and only execute if there is no error.

subash-suresh replied:
Updated the reserve session and create dcpower VIs to use error case structures as they need to be skipped when there is an error.
- Reserver Session
![image](https://github.com/ni/measurementlink-labview/assets/98541596/c130d89b-f8b4-46b8-b84a-fe4cc97ec166)
![image](https://github.com/ni/measurementlink-labview/assets/98541596/6e8751dd-a3ba-4e88-ad02-a47de2946591)

- Create DC Power Session
![image](https://github.com/ni/measurementlink-labview/assets/98541596/9ce846b4-c920-4bbe-9529-f7df43fbd60c)

- DC Power Specific Logic
![image](https://github.com/ni/measurementlink-labview/assets/98541596/235eda3d-48b6-489b-aa6a-7f6ad5a04a86)

subash-suresh commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Get Session Information.vi:
The icon has been updated to mention dcpower

subash-suresh commented:
Latest updates on 6/15/23,

@dixonjoel, @pbirkhol-ni , @bkeryan 
I have moved the logic of getting the gRPC device location out of the reserve session api into a seperate helper vi. This is because the reserve sessions API can be used for a device driver that is not part of the grpc device server based on the feedback from Brad in #207 .

This new vi is called `Get gRPC device location.vi`,
This VI receives the gRPC channel pool as input, and if the discovery service id is 0 then it trys to create a discovery service client.
![image](https://github.com/ni/measurementlink-labview/assets/98541596/49359a46-b52a-41f5-b7c6-560eb5ba4b46)
![image](https://github.com/ni/measurementlink-labview/assets/98541596/be0aec74-f01b-4d27-b8ef-304076069e72)

The `Create DC Power Sessions.vi` uses the `Get gRPC device location.vi` to get the device server location. Hence the VI has been updated to accept gRPC channel pool.
![image](https://github.com/ni/measurementlink-labview/assets/98541596/70d8f0c3-9456-4c25-98da-4f9f1f092a88)

Based on these updates the measurement logic looks like below,
![image](https://github.com/ni/measurementlink-labview/assets/98541596/67568b98-feb8-4530-be8d-3b452f31f951)

I am resetting reviewers due to this new update

pbirkhol-ni commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Reserve Sessions.vi:
In the error case, the MeasurementContext class should be wired through instead of creating a new MeausurementContext object. As a general rule, any input that has a corresponding output should be wired through the error case. This usually applies to error wires, classes, refnums, etc.

Also, this isn't strictly necessary, but for things like the `session information` array and the `gRPC channel pool out` cluster, you can just right click on the tunnel and select `Use Default if Unwired`. Then you don't have to create a constant in the Error case just to pass out default data. I think this can be clearer to someone reading the code since the tunnel changes appearance and you don't have to look at the Error case to know what is happening.

pbirkhol-ni replied:
This same comment applies to other VIs where you added the error case structure.

pbirkhol-ni replied:
Example:
![image](https://github.com/ni/measurementlink-labview/assets/57813166/7b67012f-4932-43a6-b11e-cf7380c1867b)

pbirkhol-ni commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Get Session Information.vi:
The `session information` indicator isn't wired.

subash-suresh replied:
I have wired the session information and made the input as a `required` input to avoid this in future

pbirkhol-ni commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/Measurement Logic.vi:
The Instrument Handle isn't wired in this VI.

Make sure you are testing the example before you submit. Both success and error cases. As is, it couldn't possibly work.

subash-suresh replied:
I have wired this input and made it required to avoid such errors in future. 

I have tested the VIs.

pbirkhol-ni reviewed (changes requested):
Make sure you use required inputs where necessary. As a general rule, _all_ inputs should be required unless we can come up with a sane default. So timeouts can be recommended (or even optional), but sessions, class, etc should be required.

jasonmreding commented:
I have mixed feelings on the channel pool. So it may just be a terminology/naming issue, but I don't really consider this a channel pool as it only exists for the lifetime of a single RPC on the server. Ideally, a channel pool would be utilized so that channels/clients can be reused across multiple RPCs. However, this brings in the complication of knowing when to remove invalid/stale channels from the pool because the service is no longer running on that port. 

If we're trying to avoid taking on the complexity/implementation cost of invalidating the cache for a channel pool, then maybe just tweaking the name is the way to go. However, I'm not really coming up with any good suggestions at the moment. Maybe just calling it something like GRPC Clients.ctl would be better.

If you want to take on creating a proper channel pool (which I think would be valuable), then I think wrapping it in a class or some other API (functional globals) would be better. This would allow you to remove some duplicated for instance to get the discovery client and then update the grpc id in the cluster. Ideally the channel pool would be generic such you could store any client in the pool, although maybe we add specific helpers that make it easy to get "well known" clients like session manager, discovery service, etc. Because the LV driver APIs don't have the concept of a gRPC channel in their create/init APIs, having the channel pool also take on the responsibility of remember past resolve calls to the discovery service for the host:port of the driver API is also useful. The caching of these two things (host:port returned from resolve and clients created from grpc-labview) make a huge difference in run time performance if a service is called multiple times (i.e. sweep loop) in a test sequence.

bkeryan commented:
> However, this brings in the complication of knowing when to remove invalid/stale channels from the pool because the service is no longer running on that port.

This is why our other channel pools are indexed by hostname:port, not service interface.

bkeryan commented:
> I don't think this new VI should be creating a DiscoveryClient. It should require that something external to it create the DiscoveryClient. It should not take a cluster as an input (since it doesn't care about the session manager) and instead just take the DiscoveryClient id. We should consider making a ring control with some common grpc device service classes, but also allow raw text for service classes that we don't know about yet.

This seems like a performance problem. I feel like we are going to have to start caching the discovery client and/or the gRPC channel for the discovery service.

bkeryan commented:
> I have mixed feelings on the channel pool. So it may just be a terminology/naming issue, but I don't really consider this a channel pool as it only exists for the lifetime of a single RPC on the server. Ideally, a channel pool would be utilized so that channels/clients can be reused across multiple RPCs. However, this brings in the complication of knowing when to remove invalid/stale channels from the pool because the service is no longer running on that port.
> 
> If we're trying to avoid taking on the complexity/implementation cost of invalidating the cache for a channel pool, then maybe just tweaking the name is the way to go. However, I'm not really coming up with any good suggestions at the moment. Maybe just calling it something like GRPC Clients.ctl would be better.
> 
> If you want to take on creating a proper channel pool (which I think would be valuable), then I think wrapping it in a class or some other API (functional globals) would be better. This would allow you to remove some duplicated for instance to get the discovery client and then update the grpc id in the cluster. Ideally the channel pool would be generic such you could store any client in the pool, although maybe we add specific helpers that make it easy to get "well known" clients like session manager, discovery service, etc. Because the LV driver APIs don't have the concept of a gRPC channel in their create/init APIs, having the channel pool also take on the responsibility of remember past resolve calls to the discovery service for the host:port of the driver API is also useful. The caching of these two things (host:port returned from resolve and clients created from grpc-labview) make a huge difference in run time performance if a service is called multiple times (i.e. sweep loop) in a test sequence.

I agree with pretty much all of this. It's misleading to call this cluster a channel pool, but it would be okay to call it `gRPC Clients` or something. And I feel like we will need to write an actual channel pool, sooner or later.

bkeryan commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Create DC Power Session.vi:
The driver's proper name is "NI-DCPower". It's okay to omit the dash in PascalCase strings but don't replace it with a space.

`Create NI-DCPower Session.vi`
`NI-DCPower Specific Logic.vi`
`Destroy NI-DCPower Session.vi`

subash-suresh replied:
updated the names of the VIs as requested

bkeryan commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/DCPower Specific Logic.vi:
Removing init/close from this VI makes it simpler, but now it's no longer a standalone VI that you can run directly.

subash-suresh replied:
what wpuld be the recommended approach here. Is it better to have the logic separated into different VIs

bkeryan commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Create DC Power Session.vi:
I just noticed we're not using the same terminology as the driver APIs. Should we use init/close instead of create/destroy?

subash-suresh replied:
I have updated the terminology to match the DC Power VIs i.e., I used init/close.

bkeryan commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Create DC Power Session.vi:
Can we make `Location to DCPower Grpc Device.vi` a malleable VI? I'm not sure if it would require wiring an empty `NI gRPC Device Server` as an input or whether it could adapt to the output connector type.

subash-suresh replied:
Since the input for the NI DC Power init method's input where this `Location to DCPower Grpc Device.vi` VI's  output is mapped is not a strict type def, we made the output as a simple cluster and let LabVIEW handle the reset.
![image](https://github.com/ni/measurementlink-labview/assets/98541596/556db635-7d2b-4691-9fae-7defada9a40b)
![image](https://github.com/ni/measurementlink-labview/assets/98541596/a95ee19a-9bbb-4b17-b2e3-404e586909ff)

pbirkhol-ni commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Reserve Session(s).vi:
I don't think adapting to output works, but adapting to input does. I think Reserve has to show the selector since it won't adapt properly. Unreserve will adapt, but I'm also not opposed to showing the selector. It's not like showing the selector makes it no longer adapt. In short, I don't care either way. As long as Reserve shows the selector I'm happy.

subash-suresh replied:
I have added the poly selector for unreserve as well to avoid confusion.
![image](https://github.com/ni/measurementlink-labview/assets/98541596/9cd72343-aa86-45b5-bf88-626b4a0da05d)

bkeryan commented:
> I don't think this new VI should be creating a DiscoveryClient. It should require that something external to it create the DiscoveryClient. It should not take a cluster as an input (since it doesn't care about the session manager) and instead just take the DiscoveryClient id. We should consider making a ring control with some common grpc device service classes, but also allow raw text for service classes that we don't know about yet.

I agree. It feels like this VI has too many responsibilities. 

Consider a measurement that reserves two different instrument types, like niScope and niFgen. If this measurement calls reserve twice with different instrument type ids, then it also creates two connections to the discovery service and session management service.

However, I also have a nagging feeling that performing multiple reservations is not good due to lock ordering. It may be better to omit the instrument type id in this case.

subash-suresh commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Get Session Information.vi:
My bad, i have fixed this

subash-suresh commented:
> > Latest updates on 6/15/23,
> > @dixonjoel, @pbirkhol-ni , @bkeryan I have moved the logic of getting the gRPC device location out of the reserve session api into a seperate helper vi. This is because the reserve sessions API can be used for a device driver that is not part of the grpc device server based on the feedback from Brad in #207 .
> > This new vi is called `Get gRPC device location.vi`, This VI receives the gRPC channel pool as input, and if the discovery service id is 0 then it trys to create a discovery service client. ![image](https://user-images.githubusercontent.com/98541596/246085442-49359a46-b52a-41f5-b7c6-560eb5ba4b46.png) ![image](https://user-images.githubusercontent.com/98541596/246087689-be0aec74-f01b-4d27-b8ef-304076069e72.png)
> > The `Create DC Power Sessions.vi` uses the `Get gRPC device location.vi` to get the device server location. Hence the VI has been updated to accept gRPC channel pool. ![image](https://user-images.githubusercontent.com/98541596/246087795-70d8f0c3-9456-4c25-98da-4f9f1f092a88.png)
> > Based on these updates the measurement logic looks like below, ![image](https://user-images.githubusercontent.com/98541596/246088121-67568b98-feb8-4530-be8d-3b452f31f951.png)
> > I am resetting reviewers due to this new update
> 
> I don't think this new VI should be creating a DiscoveryClient. It should require that something external to it create the DiscoveryClient. It should not take a cluster as an input (since it doesn't care about the session manager) and instead just take the DiscoveryClient id. We should consider making a ring control with some common grpc device service classes, but also allow raw text for service classes that we don't know about yet.

I have updated the new VI to take in `discovery service gRPC ID` 
![image](https://github.com/ni/measurementlink-labview/assets/98541596/98ae108f-ba05-43a7-968f-eea6c34b3cf0)

and moved the discovery service creation outside into a separate VI called `Create Discovery Client` 
![image](https://github.com/ni/measurementlink-labview/assets/98541596/8f648c07-6380-41dd-b4c4-8f2b4ca80ee2)

This VI is called in the `Initialize NI DC Power Session.vi` before calling the `Get gRPC Device Location.vi` . 
![image](https://github.com/ni/measurementlink-labview/assets/98541596/c1bb07c8-e9f5-4f67-bdbd-b42c5085771a)

subash-suresh commented:
> > I have mixed feelings on the channel pool. So it may just be a terminology/naming issue, but I don't really consider this a channel pool as it only exists for the lifetime of a single RPC on the server. Ideally, a channel pool would be utilized so that channels/clients can be reused across multiple RPCs. However, this brings in the complication of knowing when to remove invalid/stale channels from the pool because the service is no longer running on that port.
> > If we're trying to avoid taking on the complexity/implementation cost of invalidating the cache for a channel pool, then maybe just tweaking the name is the way to go. However, I'm not really coming up with any good suggestions at the moment. Maybe just calling it something like GRPC Clients.ctl would be better.
> > If you want to take on creating a proper channel pool (which I think would be valuable), then I think wrapping it in a class or some other API (functional globals) would be better. This would allow you to remove some duplicated for instance to get the discovery client and then update the grpc id in the cluster. Ideally the channel pool would be generic such you could store any client in the pool, although maybe we add specific helpers that make it easy to get "well known" clients like session manager, discovery service, etc. Because the LV driver APIs don't have the concept of a gRPC channel in their create/init APIs, having the channel pool also take on the responsibility of remember past resolve calls to the discovery service for the host:port of the driver API is also useful. The caching of these two things (host:port returned from resolve and clients created from grpc-labview) make a huge difference in run time performance if a service is called multiple times (i.e. sweep loop) in a test sequence.
> 
> I agree with pretty much all of this. It's misleading to call this cluster a channel pool, but it would be okay to call it `gRPC Clients` or something. And I feel like we will need to write an actual channel pool, sooner or later.

I have renamed the control to `GRPC Clients.ctl` for now and we will plan on taking the gRPC channel implementation in a later PR as suggested.

subash-suresh commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/Measurement Logic.vi:
I have updated the gRPC channel pool output to be a side connector and the updated block diagram looks like below
![image](https://github.com/ni/measurementlink-labview/assets/98541596/7da5c286-06f0-43b8-beeb-1045b748c1f5)

![image](https://github.com/ni/measurementlink-labview/assets/98541596/dfc451b2-5baf-4d5d-b47a-14c82d5b71ad)

subash-suresh commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Reserve Session.vi:
udpated the VIs to be part of the library

subash-suresh commented:
> > I don't think this new VI should be creating a DiscoveryClient. It should require that something external to it create the DiscoveryClient. It should not take a cluster as an input (since it doesn't care about the session manager) and instead just take the DiscoveryClient id. We should consider making a ring control with some common grpc device service classes, but also allow raw text for service classes that we don't know about yet.
> 
> I agree. It feels like this VI has too many responsibilities.
> 
> Consider a measurement that reserves two different instrument types, like niScope and niFgen. If this measurement calls reserve twice with different instrument type ids, then it also creates two connections to the discovery service and session management service.
> 
> However, I also have a nagging feeling that performing multiple reservations is not good due to lock ordering. It may be better to omit the instrument type id in this case. Otherwise, if some measurements reserve niScope then niFgen and others reserve niFgen then niScope, they may deadlock (until the reservation timeout).



> > I don't think this new VI should be creating a DiscoveryClient. It should require that something external to it create the DiscoveryClient. It should not take a cluster as an input (since it doesn't care about the session manager) and instead just take the DiscoveryClient id. We should consider making a ring control with some common grpc device service classes, but also allow raw text for service classes that we don't know about yet.
> 
> I agree. It feels like this VI has too many responsibilities.
> 
> Consider a measurement that reserves two different instrument types, like niScope and niFgen. If this measurement calls reserve twice with different instrument type ids, then it also creates two connections to the discovery service and session management service.
> 
> However, I also have a nagging feeling that performing multiple reservations is not good due to lock ordering. It may be better to omit the instrument type id in this case. Otherwise, if some measurements reserve niScope then niFgen and others reserve niFgen then niScope, they may deadlock (until the reservation timeout).

I have made the instrument type input as optional input and if the users do not give the instrument type input it automatically reserves all the session without filtering based on instrument type. @bkeryan

pbirkhol-ni commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Initialize NI-DCPower Session.vi:
This still isn't right. You should not be creating a DiscoveryClient here. You've already done that and the gRPC ID is part of the `gRPC Clients` input. If this VI were designed to be called in isolation then you'd have to create a DiscoveryClient. But it's designed to be called after reserving sessions, which will already create the DiscoveryClient. 

![image](https://github.com/ni/measurementlink-labview/assets/57813166/acd0676a-26f2-4f96-a6b1-b807fa611921)

pbirkhol-ni replied:
Also, because it isn't designed to be use in isolation, the `gRPC Clients` input should be required.

subash-suresh replied:
updated as per suggestion

pbirkhol-ni commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/NI-DCPower Specific Logic.vi:
Use the default if unwired option in the error case.
![image](https://github.com/ni/measurementlink-labview/assets/57813166/e15e5adf-7701-488f-9481-0c5cb35e9a22)

subash-suresh replied:
updated as suggested.

subash-suresh commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Reserve (N Sessions).vi:
Updated the logic to take the ds client id from resolve session call.
![image](https://github.com/ni/measurementlink-labview/assets/98541596/a092863d-bd47-449d-b4b0-c267c7326092)

subash-suresh commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Reserve (N Sessions).vi:
removed the unwanted error wire

subash-suresh commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Reserve (1 Session).vi:
Updated all the places where gRPC channel pool was mentioned

subash-suresh commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/GRPC Clients.ctl:
updated to `gRPC clients`

subash-suresh replied:
![Uploading image.png…]()

pbirkhol-ni reviewed (approved):
I think this meets the requirements of the PO. I'm not super thrilled with it, but it doesn't commit us to anything long term, so we can always change it later.

subash-suresh commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Initialize NI-DCPower Session.vi:
udpated the logic to remove rebundling 
![Uploading image.png…]()

subash-suresh commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Initialize NI-DCPower Session.vi:
updated the comments

subash-suresh commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Get gRPC Device Location.vi:
updated to use empty stirngs by using the default cluster value
![image](https://github.com/ni/measurementlink-labview/assets/98541596/613d5b38-18a0-49b8-93e7-818ecd15ac11)

subash-suresh commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Get gRPC Device Location.vi:
marked the `discovery service gRPC ID` as required.
![image](https://github.com/ni/measurementlink-labview/assets/98541596/c46c50c4-3b05-499f-ac2b-4d9c7a94280c)

subash-suresh commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Reserve (1 Session).vi:
marked the error in as recommended as suggested.
![image](https://github.com/ni/measurementlink-labview/assets/98541596/0daa37b9-841b-4a87-8f92-eb25fd61ac4d)

jasonmreding reviewed (approved):
My only comment is that it might be worth trying to extract driver agnostic code into a shared lvlib at some point that is referenced by all our examples. That doesn't need to be done now though. The library could go at the root of the "Example Measurements" folder. However, I don't know how well LV does with referencing files relatively that aren't under or in the same directory as the project file. Even collecting some driver specific APIs in this library would probably be OK. This could be treated as a preview/alpha version of what might become an official public API in vilib at some point.

bkeryan commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Reserve Session(s).vi:
You capitalized the 'p' in "poly" but I guess that's okay. I checked codesearch and there is no consistent standard for "(poly).vi" vs. "(Poly).vi"

bkeryan commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Unreserve (Poly).vi:
Uncheck "Show Selector by Default"
![image](https://github.com/ni/measurementlink-labview/assets/17914399/3b2be956-2311-4cf1-b246-c4cf3e40ee26)

subash-suresh replied:
updated as suggested

bkeryan replied:
Why did you uncheck it for Reserve?

bkeryan replied:
`Measurement Logic.vi` shows the selector for `Reserve` and hides the selector for `Unreserve`. These should be the defaults in the poly VIs.

subash-suresh replied:
I have updated the reserve sesison to have it checked and unreserved to have this unchecked as that is the expected default value. This update can be observed in #230

bkeryan commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/NI-DCPower Specific Logic.vi:
Select the non-error case and resave

subash-suresh replied:
updated as suggested

bkeryan reviewed (approved):
Approved with suggestions

bkeryan commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Initialize NI-DCPower Session.vi:
> Creates and returns an gRPC based NI DC power instrument handle. 

Only use "an" when the following word starts with a vowel.

subash-suresh replied:
This has been handled in #230

bkeryan commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Initialize NI-DCPower Session.vi:
> Creates and returns an gRPC based NI DC power instrument handle. 

branding: "NI-DCPower"

subash-suresh replied:
This has been updated in #230

bkeryan commented on Source/Example Measurements/NI-DCPower Source DC Voltage/NI-DCPower Measurement/subVIs/Close NI-DCPower Session.vi:
> Closes the given NI DC power instrument handle.

branding: "NI-DCPower"

subash-suresh replied:
This has been resolved in #230

subash-suresh commented:
I will make the suggested documentation changes and poly VI setting changes in a new PR.
